library/core/src/pin.rs

```diff
@@ -2014,5 +2019,145 @@ pub macro pin($value:expr $(,)?) {
     //
     // See https://doc.rust-lang.org/1.58.1/reference/destructors.html#temporary-lifetime-extension
     // for more info.
-    $crate::pin::Pin::<&mut _> { __pointer: &mut { $value } }
+    $crate::pin::Pin::<&mut _> {
+        __pointer: &mut $crate::pin::__Pinned {
+            value_expression_preserving_lifespan_extension_of_temporaries: $value,
+        } as &mut _,
+        // ^^^^^^^^
+        // DerefMut coërcion, because of --+
+        __phantom: { // <------------------+
+            let pin = $crate::marker::PhantomData;
+            if false {
+                loop {}
+                // We use dead code to disable move semantics (and borrowck) to allow us to refer to
+                // `$value` again even when `$value` happens to consume non-`Copy` stuff or whatnot.
+                #[allow(unreachable_code)] {
+                    $crate::pin::__phantomdata_set_typeof_pointee(pin, $value);
+                }
+            }
+            pin
+        },
+    } as $crate::pin::Pin<&mut _> // allow unsized coërcions
 }
+
+/// Since the 2024 edition, the rules for lifespan-of-temporaries extension have changed,
+/// and `{ $value }` no longer cuts it for all the scenarios where it did in the 2021 edition and
+/// before.
+///
+///   - aside: throughout this documentation the expression "lifespan extension" shall be preferred
+///     to that of "lifetime extension", so as to clarify we are firstly talking of the lifespan of
+///     a temporary, which, when extended, results in the lifetime of a borrow to it to be allowed
+///     to be bigger. This nit seems more in line with the idea that lifetimes themselves are just
+///     descriptive, not prescriptive, so it does not make that much sense to talk of extending one
+///     _directly_. We would not need to lift this ambiguity had Rust chosen a term other than
+///     "lifetime" to talk of the duration of borrows.
+///
+/// Indeed, things such as the following regressed:
+///
+/// ```rust
+/// // Regression test for #138596
+///
+/// fn main() {
+///     match core::pin::pin!(foo(&mut 0)) {
+///         _f => {}
+///     }
+/// }
+///
+/// async fn foo(_s: &mut usize) {}
+/// ```
+///
+/// The reason for that is that we'd end up with `{ foo(&mut temporary(0)) }` among the expansion
+/// of `pin!`, and since this happens in `match`-scrutinee position, temporaries get not to outlive
+/// their encompassing braced block(s), so `temporary(0)` dies before that `}`, and the resulting
+/// `Pin { … }` is born dangling.
```

Comment:
It doesn't matter about this being in `match` scrutinee position.  E.g., this also fails in Rust 2024 and works in Rust 2021:

```rust
fn f<'a, T>(x: &'a T) -> &'a T { x }
fn g<T>(_: T) {}

struct W<T: ?Sized> {
    _inner: T,
}

fn main() {
    _ = g(W { _inner: &mut { f(&mut ()) } });
}
```